ethers-solc/src/cache.rs

```diff
@@ -25,11 +26,34 @@ use std::{
 /// `ethers-solc` uses a different format version id, but the actual format is consistent with
 /// hardhat This allows ethers-solc to detect if the cache file was written by hardhat or
 /// `ethers-solc`
-const ETHERS_FORMAT_VERSION: &str = "ethers-rs-sol-cache-3";
+const ETHERS_FORMAT_VERSION: &str = "ethers-rs-sol-cache-4";
 
 /// The file name of the default cache file
 pub const SOLIDITY_FILES_CACHE_FILENAME: &str = "solidity-files-cache.json";
 
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub struct CompilationUnitId(String);
+
+impl CompilationUnitId {
+    /// Create a unique id for each compilation unit based on compiler version and settings
+    pub fn new(version: Version, solc_config: SolcConfig) -> Self {
+        let mut hasher = hash_map::DefaultHasher::new();
+        version.hash(&mut hasher);
+        solc_config.hash(&mut hasher);
+        let result = hasher.finish();
+        Self(format!("{result:x}"))
+    }
+}
+
+/// A unit of source files that are compiled together with the same compiler version and settings
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CompilationUnit {
+    pub solc_config: SolcConfig,
+    pub version: Version,
+    pub source_units: HashSet<PathBuf>,
```

Comment:
let's make this ordered?

```suggestion
    pub source_units: BTreeMap<PathBuf>,
```

Reply:
this is written to disk, so we want consistent order